### PR TITLE
feat: pas-446 create a vds subsetting script

### PR DESCRIPTION
**Pull request by Alexander-Stuckey**
Create a script that will take in a VDS path, an output format (although only VDS works right now) and optionally a number of samples and / or list of regions to subset down to.

The script will then subset down the VDS to those samples / regions, and output the requested output files + a sample list (if subset) to GCS.

### CODE REVIEW

**Comment by annacuomo** [comment]
looks good! Would be cool to have the option to subset to a whole chromosome rather than an interval if possible? Or have a default interval?

**Comment by Alexander-Stuckey** [comment]
> Also for SAIGE-QTL specifically, the index file associated with a VCF should be `.csi` rather than `.tbi` (tabix)!

Hail `export_vcf` doesn't support writing a `.csi` index 😞 . I figured that an index is better than no index, although it does mean that you'd have to generate your own.

File: subset_vds.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+        if not fullmatch("^[0-9chrxymt]+", interval.split(":")[0].lower()):
+            problem_intervals.append(interval)
+    if len(problem_intervals) > 0:
+        raise ValueError(
+            f"The list of supplied intervals contains the following incorrectly formatted contigs: {[problem for problem in problem_intervals]}"
+        )
+    return interval_list
+
+
+def convert_intervals_to_locus(interval_list: list[str]) -> list[IntervalExpression]:
```
Comment (MattWellie) [inline subset_vds.py:101]: I think a lot of this code block/validation you would get for free with just calling `parse_locus_interval` - it tolerates `chr`, `chr:start-stop`, `chr1-chr2`, and would give you positional validation based on the reference.

If this was partially taken from [here](https://github.com/populationgenomics/analysis-runner/blob/main/scripts/subset_matrix_table.py#L150), then yes, I probably also wasted some time.

One piece of validation you might be missing here is checking for a `chr` prefix on the chromosome name, or a check that the chromosome name appears in the hail.reference.contigs

File: subset_vds.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            outdata.close()
+
+
+def main(
+    vds_path: str,
+    n_samples: int | None,
+    intervals: str | None,
+    output_formats: list[str],
+) -> None:
+    allowed_output_formats: list[str] = ["vcf", "bed", "vds"]
```
Comment (MattWellie) [inline subset_vds.py:297]: you can put this validation in the argparse interface
https://docs.python.org/3/library/argparse.html#choices

Comment (MattWellie) [reply subset_vds.py:297]: wait, no you can't... not if you tolerate multiple. Curses.

File: subset_vds.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        subset_vds.write(output_path(f"{infile_name}_subset", test=is_test))
+
+    subset_dense_mt: MatrixTable | Table | Any = to_dense_mt(subset_vds)
+    subset_dense_mt = split_multi_hts(subset_dense_mt)
+
+    for format in output_formats:
+        if format == "bed":
+            export_plink(
+                subset_dense_mt,
+                output_path(f"{infile_name}_subset", test=is_test),
```
Comment (MattWellie) [inline subset_vds.py:263]: `output_path` will handle this automatically - output_path builds an output location based on a few things in config:
1. default bucket for the dataset
2. output_prefix
3. file path/suffix

[output_path](https://github.com/populationgenomics/cpg-utils/blob/main/cpg_utils/config.py#L421) calls [dataset_path](https://github.com/populationgenomics/cpg-utils/blob/main/cpg_utils/config.py#L297), which builds the full path.

For a test run, the only paths in config are to test buckets, so it will only know how to build test output paths.

For main/full runs, there will be both test and main bucket paths in config. This `test` toggle is designed so that on a full/main permission run you can choose to direct the output to a test bucket anyway.

From the way you're using this (based on the access level), you'd be directing the output to the location it would be writing to by default, so the test argument doesn't really help you here.

File: subset_vds.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    n_total_samples: int = len(starting_samples)
+    if n_samples == 0:
+        raise ValueError(
+            "Subsetting to 0 samples will result in an empty MatrixTable, so not doing that."
+        )
+    if n_samples >= n_total_samples:
+        raise ValueError(
+            f"The subset sample size {n_samples} is at least the same size as the total number of samples in the input {n_total_samples}. Please choose a smaller number."
+        )
+    subset_sample_list = sample(starting_samples, n_samples)
```
Comment (MattWellie) [inline subset_vds.py:185]: do we want to seed this for reproducibility?

**Review by MattWellie — CHANGES_REQUESTED** [review]
I think the `test=` specifier on output_path is redundant, but otherwise looks pretty hot

File: subset_vds.py (hunk 2 above)
Comment (MattWellie) [reply subset_vds.py:297]: ooooh, or can you. You could have a list of choices, set `nargs='+'` on the argument, and take multiple individual arguments. that would remove the need for 
```output_formats=args.output_formats.lower().split(",")```

Comment (Alexander-Stuckey) [reply subset_vds.py:297]: Ah, that was what I was wanting to do, I assume then it's just multiple space separated arguments to --output-formats?

File: subset_vds.py (hunk 3 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:263]: Makes sense, I can remove this

File: subset_vds.py (hunk 4 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:185]: We can get this for free, so why not

File: subset_vds.py (hunk 1 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:101]: I did use that as a reference, yes. Not familiar with using hail locuses yet.
If I can delete all this checking code and basically rely on `parse_locus_interval` to do all the checking for me, then I'm happy to do that. Thoughts?

Comment (MattWellie) [reply subset_vds.py:102]: It looks like you've already moved in that direction. Sounds good to me - if you just force a user to provide one of the Hail supported input formats, and if there are any invalid ones print a link to the Hail docs on which ones are valid, then it's still pretty flexible, and takes the burden off you writing the validation

File: subset_vds.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        "--random-seed",
+        help="Seed for the random number generator when subsetting by samples",
+        required=False,
+        default=19700101,
+        type=int,
+    )
+    args, failures = parser.parse_known_args()
+    if failures:
+        parser.print_help()
+        raise AttributeError(f"Invalid arguments {failures}")
```
Comment (MattWellie) [inline subset_vds.py:329]: It's not important here, but I completely misunderstood this method, and was using it incorrectly.

`parser.parse_args()` attempts to parse all arguments, and will fail if there are additional arguments provided.

`parser.parse_known_args()` only parses the listed arguments, and will tolerate other arguments being present (it'll return the other un-digested ones). I'm not sure when you'd want to parse only a subset of the CLI arguments, but that would be permitted in parse_known_args. 

TL;DR - if you always want to fail when unknown arguments are present, the standard parse_args is the way to go

File: subset_vds.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        if to_path(output_path("subset_samples_file.txt")).exists():
+            raise FileExistsError(
+                f"The file {to_path(output_path('subset_samples_file.txt'))} exists. Refusing to overwrite it."
+            )
+        outdata = StringIO()
+        for single_sample in subset_sample_list:
+            outdata.write(f"{single_sample}\n")
+        with to_path(output_path("subset_samples_file.txt")).open("wt") as outfile:
+            outfile.write(outdata.getvalue())
+            outdata.close()
```
Comment (MattWellie) [inline subset_vds.py:258]:
```suggestion
        with to_path(output_path("subset_samples_file.txt")).open("wt") as outfile:
            outfile.write("\n".join(subset_sample_list))
```

You should just be able to write the string content directly without the StringIO intermediary

File: subset_vds.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+            if "gvcf_info" in subset_dense_mt.entry:
+                subset_dense_mt = subset_dense_mt.drop("gvcf_info")
+            export_vcf(
+                subset_dense_mt,
+                output_path(f"{infile_name}_subset.vcf.bgz"),
+                tabix=True,
+            )
+
+    if subset_sample_list:
+        if to_path(output_path("subset_samples_file.txt")).exists():
```
Comment (MattWellie) [inline subset_vds.py:249]: This is a bit of a nit, but you might want to move this existence check earlier in the process, else you could potentially be in a position where the subset itself has been written, but you raise an exception before writing the sample list.

That's probably not going to happen, and you can pull the Sample subset easily from the VDS itself, so not a real issue

File: subset_vds.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+        ValueError: if n_samples is >= to the number of samples in the input vds throw an error, as this would not do any filtering
+
+    Returns:
+        list[str]: a random list of samples of length n_samples from the input vds, to use in subsetting
+    """
+    starting_samples: list[str] = input_vds.variant_data.s.collect()
+    n_total_samples: int = len(starting_samples)
+    if n_samples == 0:
+        raise ValueError(
+            "Subsetting to 0 samples will result in an empty MatrixTable, so not doing that."
```
Comment (MattWellie) [inline subset_vds.py:156]: > so not doing that

yeah, you tell 'em

**Review by MattWellie — APPROVED** [review]

File: subset_vds.py (hunk 6 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:258]: I can, this just avoids multiple writes (one per line) to GCP, instead writing the whole file at once. I figured this'd be better practice

File: subset_vds.py (hunk 5 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:329]: Ah, I had assumed that the functionality was reversed, as I'd seen it used around CPG like that. I'll make that change

File: subset_vds.py (hunk 7 above)
Comment (Alexander-Stuckey) [reply subset_vds.py:249]: Fair point though, changed that

File: subset_vds.py (hunk 5 above)
Comment (MattWellie) [reply subset_vds.py]: Almost certainly me writing it/others copying my mistake 😆

**Review by MattWellie — APPROVED** [review]
